File: lib/tauri/window.ts
import { invoke } from '@tauri-apps/api/core';

/**
 * Window management functions
 */
export const windowCommands = {
  /**
   * Minimize the application window
   */
  minimize: async (): Promise<void> => {
    try {
      await invoke('minimize_window');
    } catch (error) {
      throw new Error(`Failed to minimize window: ${error}`);
    }
  },

  /**
   * Maximize the application window
   */
  maximize: async (): Promise<void> => {
    try {
      await invoke('maximize_window');
    } catch (error) {
      throw new Error(`Failed to maximize window: ${error}`);
    }
  },

  /**
   * Unmaximize (restore) the application window
   */
  unmaximize: async (): Promise<void> => {
    try {
      await invoke('unmaximize_window');
    } catch (error) {
      throw new Error(`Failed to unmaximize window: ${error}`);
    }
  },

  /**
   * Close the application window
   */
  close: async (): Promise<void> => {
    try {
      await invoke('close_window');
    } catch (error) {
      throw new Error(`Failed to close window: ${error}`);
    }
  },

  /**
   * Check if window is maximized
   */
  isMaximized: async (): Promise<boolean> => {
    try {
      return await invoke<boolean>('is_window_maximized');
    } catch (error) {
      throw new Error(`Failed to check window state: ${error}`);
    }
  },

  /**
   * Toggle fullscreen mode
   */
  toggleFullscreen: async (): Promise<void> => {
    try {
      await invoke('toggle_fullscreen');
    } catch (error) {
      throw new Error(`Failed to toggle fullscreen: ${error}`);
    }
  },

  /**
   * Set window title
   */
  setTitle: async (title: string): Promise<void> => {
    try {
      await invoke('set_window_title', { title });
    } catch (error) {
      throw new Error(`Failed to set window title: ${error}`);
    }
  },

  /**
   * Get window size
   */
  getSize: async (): Promise<{ width: number; height: number }> => {
    try {
      const [width, height] = await invoke<[number, number]>('get_window_size');
      return { width, height };
    } catch (error) {
      throw new Error(`Failed to get window size: ${error}`);
    }
  },

  /**
   * Set window size
   */
  setSize: async (width: number, height: number): Promise<void> => {
    try {
      await invoke('set_window_size', { width, height });
    } catch (error) {
      throw new Error(`Failed to set window size: ${error}`);
    }
  },

  /**
   * Center window on screen
   */
  center: async (): Promise<void> => {
    try {
      await invoke('center_window');
    } catch (error) {
      throw new Error(`Failed to center window: ${error}`);
    }
  },

  /**
   * Show window
   */
  show: async (): Promise<void> => {
    try {
      await invoke('show_window');
    } catch (error) {
      throw new Error(`Failed to show window: ${error}`);
    }
  },

  /**
   * Hide window
   */
  hide: async (): Promise<void> => {
    try {
      await invoke('hide_window');
    } catch (error) {
      throw new Error(`Failed to hide window: ${error}`);
    }
  },
};

File: src-tauri/src/commands/window.rs
use tauri::Window;

#[tauri::command]
pub async fn minimize_window(window: Window) -> Result<(), String> {
    window
        .minimize()
        .map_err(|e| format!("Failed to minimize window: {}", e))?;
    Ok(())
}

#[tauri::command]
pub async fn maximize_window(window: Window) -> Result<(), String> {
    window
        .maximize()
        .map_err(|e| format!("Failed to maximize window: {}", e))?;
    Ok(())
}

#[tauri::command]
pub async fn unmaximize_window(window: Window) -> Result<(), String> {
    window
        .unmaximize()
        .map_err(|e| format!("Failed to unmaximize window: {}", e))?;
    Ok(())
}

#[tauri::command]
pub async fn close_window(window: Window) -> Result<(), String> {
    window
        .close()
        .map_err(|e| format!("Failed to close window: {}", e))?;
    Ok(())
}

#[tauri::command]
pub async fn is_window_maximized(window: Window) -> Result<bool, String> {
    window
        .is_maximized()
        .map_err(|e| format!("Failed to check window state: {}", e))
}

#[tauri::command]
pub async fn toggle_fullscreen(window: Window) -> Result<(), String> {
    let is_fullscreen = window
        .is_fullscreen()
        .map_err(|e| format!("Failed to check fullscreen state: {}", e))?;
    
    window
        .set_fullscreen(!is_fullscreen)
        .map_err(|e| format!("Failed to toggle fullscreen: {}", e))?;
    
    Ok(())
}

#[tauri::command]
pub async fn set_window_title(window: Window, title: String) -> Result<(), String> {
    window
        .set_title(&title)
        .map_err(|e| format!("Failed to set window title: {}", e))?;
    Ok(())
}

#[tauri::command]
pub async fn get_window_size(window: Window) -> Result<(u32, u32), String> {
    let size = window
        .inner_size()
        .map_err(|e| format!("Failed to get window size: {}", e))?;
    
    Ok((size.width, size.height))
}

#[tauri::command]
pub async fn set_window_size(window: Window, width: u32, height: u32) -> Result<(), String> {
    use tauri::Size;
    
    window
        .set_size(Size::Physical(tauri::PhysicalSize { width, height }))
        .map_err(|e| format!("Failed to set window size: {}", e))?;
    
    Ok(())
}

#[tauri::command]
pub async fn center_window(window: Window) -> Result<(), String> {
    window
        .center()
        .map_err(|e| format!("Failed to center window: {}", e))?;
    Ok(())
}

#[tauri::command]
pub async fn show_window(window: Window) -> Result<(), String> {
    window
        .show()
        .map_err(|e| format!("Failed to show window: {}", e))?;
    Ok(())
}

#[tauri::command]
pub async fn hide_window(window: Window) -> Result<(), String> {
    window
        .hide()
        .map_err(|e| format!("Failed to hide window: {}", e))?;
    Ok(())
}

/// Opens a URL in the system's default browser
/// This bypasses iframe restrictions (X-Frame-Options) that block sites like google.com
#[tauri::command]
#[allow(deprecated)]
pub async fn open_external_url(app: tauri::AppHandle, url: String) -> Result<(), String> {
    use tauri_plugin_shell::ShellExt;
    
    // Validate URL format
    if !url.starts_with("http://") && !url.starts_with("https://") {
        return Err("Invalid URL: must start with http:// or https://".to_string());
    }
    
    app.shell()
        .open(&url, None)
        .map_err(|e| format!("Failed to open URL in browser: {}", e))?;
    
    Ok(())
}
